DOC: Examples for DataFrame.from_records

bdice opened the pull request:
This PR adds examples to the documentation for `pandas.DataFrame.from_records`. It also rewords a couple sentences to clarify that the input data can be a sequence of dicts or sequence of tuples (but not a sequence of DataFrames).

- [x] closes #34057
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

bdice committed:
Added examples to DataFrame.from_records.

bdice committed:
Clarify allowable inputs to from_records.

bdice committed:
Allow long type in flake8.

bdice commented:
Validated docstring, pasting here as requested by [the contributing guide](https://pandas.pydata.org/docs/dev/development/contributing.html?highlight=whatsnew#updating-a-pandas-docstring).
```python
$ python scripts/validate_docstrings.py pandas.DataFrame.from_records

################################################################################
################## Docstring (pandas.DataFrame.from_records)  ##################
################################################################################

Convert structured or record ndarray to DataFrame.

Creates a DataFrame object from a structured ndarray, sequence of
tuples, sequence of dicts, or DataFrame.

Parameters
----------
data : ndarray (structured dtype), sequence of tuples, sequence of dicts, or DataFrame
    Structured input data.
index : str, list of fields, array-like
    Field of array to use as the index, alternately a specific set of
    input labels to use.
exclude : sequence, default None
    Columns or fields to exclude.
columns : sequence, default None
    Column names to use. If the passed data do not have names
    associated with them, this argument provides names for the
    columns. Otherwise this argument indicates the order of the columns
    in the result (any names not found in the data will become all-NA
    columns).
coerce_float : bool, default False
    Attempt to convert values of non-string, non-numeric objects (like
    decimal.Decimal) to floating point, useful for SQL result sets.
nrows : int, default None
    Number of rows to read if data is an iterator.

Returns
-------
DataFrame

See Also
--------
DataFrame.from_dict : DataFrame from dict of array-like or dicts.
DataFrame : DataFrame object creation using constructor.

Examples
--------
Data can be provided as a structured ndarray:

>>> data = np.array([(3, 'a'), (2, 'b'), (1, 'c'), (0, 'd')],
...                 dtype=[('col_1', 'i4'), ('col_2', 'U1')])
>>> pd.DataFrame.from_records(data)
   col_1 col_2
0      3     a
1      2     b
2      1     c
3      0     d

Data can be provided as a list of dicts:

>>> data = [{'col_1': 3, 'col_2': 'a'},
...         {'col_1': 2, 'col_2': 'b'},
...         {'col_1': 1, 'col_2': 'c'},
...         {'col_1': 0, 'col_2': 'd'}]
>>> pd.DataFrame.from_records(data)
   col_1 col_2
0      3     a
1      2     b
2      1     c
3      0     d

Data can be provided as a list of tuples with corresponding columns:

>>> data = [(3, 'a'), (2, 'b'), (1, 'c'), (0, 'd')]
>>> pd.DataFrame.from_records(data, columns=['col_1', 'col_2'])
   col_1 col_2
0      3     a
1      2     b
2      1     c
3      0     d

################################################################################
################################## Validation ##################################
################################################################################
```

bdice commented on pandas/core/frame.py:
```diff
@@ -1633,9 +1633,13 @@ def from_records(
         Parameters
         ----------
-        data : ndarray (structured dtype), list of tuples, dict, or DataFrame
+        data : ndarray (structured dtype), sequence of tuples, sequence of dicts, or DataFrame
```
Reviewers: this line is over the length limit. I tried breaking the line and indenting the second part of the line in a few different ways, but it caused the docstrings to render incorrectly. I had to add `noqa: E501` below so that this docstring can exceed the line length limit. If you have a better solution, please suggest it!

MarcoGorelli replied:
How about

```suggestion
        data : structured ndarray, sequence of tuples or dicts, or DataFrame
```

MarcoGorelli reviewed (changes requested):
Thanks @bdice 

Just suggesting a change on the param description

bdice committed:
Simplify wording.

Replaced "sequence of tuples, sequence of dicts" with "sequence of tuples or dicts."

bdice commented:
@MarcoGorelli Thanks for the feedback -- I considered that and wasn't sure if it would be clear, but I agree with your suggestion and have applied that change. That reduces the line length enough that the `# noqa: E501` should no longer be necessary (85 char).

bdice committed:
Use "structured ndarray" as type description.

pep8speaks commented:
Hello @bdice! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/e2a86f5ffe4bc3eaf3232e877ced62dfc5a102cf/pandas/core/frame.py):

> [Line 1237:1](https://github.com/pandas-dev/pandas/blob/e2a86f5ffe4bc3eaf3232e877ced62dfc5a102cf/pandas/core/frame.py#L1237): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 1237:1](https://github.com/pandas-dev/pandas/blob/e2a86f5ffe4bc3eaf3232e877ced62dfc5a102cf/pandas/core/frame.py#L1237): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs

bdice committed:
Fix tabs/spaces from editor misconfiguration.

jorisvandenbossche reviewed (approved):
Looks good to me!

simonjayhawkins commented:
test failures unrelated.

simonjayhawkins commented:
Thanks @bdice
